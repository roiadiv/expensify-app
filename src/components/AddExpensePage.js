import React from 'react';//using jsx
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component{    
    onSubmit = (expense)=>{
        this.props.addExpense(expense);
        this.props.history.push('/');//redirect by using browser ruoting
    }
    render(){
        return (
            <div>
            <h1>Add Expense</h1>
            <ExpenseForm 
             onSubmit = {this.onSubmit}
            />
        </div>   
        );
    }
}

// const AddExpensePage = (props)=>(
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm 
//         onSubmit = {(expense)=>{
//             props.onSubmit(expense);
//             props.history.push('/');//redirect by using browser ruoting
//         }}
//         />
//     </div>   
// ); 

//as a same we have map for state, we have one for dispatch
const mapDispatchToProps = (dispatch)=>({
    addExpense:(expense)=> dispatch(addExpense(expense))   
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);