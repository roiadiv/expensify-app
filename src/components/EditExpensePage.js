import React from 'react';//using jsx
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense,startRemoveExpense} from '../actions/expenses';


export class EditExpensePage extends React.Component{
    //edit
    onSubmit = (expense)=>{
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    //remove
    onRemove = ()=>{
        this.props.startRemoveExpense({id : this.props.expense.id});
        this.props.history.push('/');
    }
    render(){
    return(
    <div>
        <ExpenseForm 
            expense = {this.props.expense}
            onSubmit = {this.onSubmit}
        />
        <button onClick = {this.onRemove}
        >Remove
        </button>
    </div>        
    )}

}

// const EditExpensePage = (props)=>(
//     <div>
//         <ExpenseForm 
//             expense = {props.expense}
//             onSubmit = {(expense)=>{
//                 props.dispatch(editExpense(props.expense.id, expense));
//                 props.history.push('/');
//             }}
//         />
//         <button onClick = {()=>{
//              props.dispatch(removeExpense({id : props.expense.id}));
//              props.history.push('/');
//             }}
//         >Remove
//     </button>
//     </div>
// );

const mapStateToProps = (state, props)=>{
    return {
        expense :  state.expenses.find((expense)=> expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch,props)=>({
    editExpense:(id,expense)=> dispatch(editExpense(id,expense)),
    startRemoveExpense:(data)=> dispatch(startRemoveExpense(data))      
});

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);