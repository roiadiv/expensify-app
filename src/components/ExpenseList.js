import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectorExpense from '../selectors/expenses';

//This component dosnt neew any more to worry or storge a state,
//He rerender by using redux store
//all we need to do to is decide how we want to render this compoennet to that screen
export const ExpenseList = (props)=>(
    <div className = "content-container">
        <div className= "list-header">
          <div className = "show-for-mobile">Expenses</div>
          <div className = "show-for-desktop">Expens</div>
          <div className = "show-for-desktop">Amount</div>
        </div>
        <div className = "list-budy">
        {
            props.expenses.length === 0 ? (
                <div class="list-item list-item--massage">
                    <span>No expenses</span>
                </div>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key = {expense.id} {...expense} />
                })
            )
        }
        </div>
    </div>   
);

const mapStateToProps = (state)=>{
    return {
        expenses :  selectorExpense(state.expenses,state.filters)
    };
};

//this is simmler to what we do in hoc.js
//the function connect have a 2 parts:
//1 - the information and the items we accses to this component from the store!
//this part get a function that have the state of the store
//this function return an object = the props we want from the store
//2 - the component itself we want to upgrade and warrep;
export default connect(mapStateToProps)(ExpenseList);
//The ConnectedExpenseList;