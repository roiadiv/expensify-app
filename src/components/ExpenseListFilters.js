import React from 'react';
import filters from '../reducers/filters';
import {connect} from 'react-redux';
import {setTextFilter, sortByAmount, sortByDate} from '../actions/filters';

const ExpenseListFilters = (props)=>(
    <div>
        <input type = "text"  value ={props.filters.text} onChange = {(e)=>{
            props.dispatch(setTextFilter(e.target.value));
        }} />

        <select value = {props.filters.sortBy} onChange = {(e)=>{
            if(e.target.value === 'date'){
                props.dispatch(sortByDate())
            }else if(e.target.value === 'amount'){
                props.dispatch(sortByAmount())
            }
        }}>
            <option value = "date">Date</option>
            <option value = "amount">Amount</option>
        </select>
    </div>

);

const mapSateToProps = (state)=>{
    return {
        filters : state.filters
    }
};

//we give mapSateToProps to use the function dispatch!!!
export default connect(mapSateToProps)(ExpenseListFilters);