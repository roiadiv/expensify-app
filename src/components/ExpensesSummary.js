import React from 'react';
import {connect} from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import selectorExpense from '../selectors/expenses';
import numeral from 'numeral';


export const ExpensesSummary = ({expensesCount,expensesTotal})=>{
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal/100).format('$0,0.00');
    return(
    <div>
           <h1>Viewing {expensesCount} {expenseWord} totalling {formattedExpensesTotal}</h1>
    </div> 
    );  
};

const mapStateToProps = (state)=>{
    const visibleExpenses = selectorExpense(state.expenses,state.filters);
    return {
        expensesCount :  visibleExpenses.length,
        expensesTotal :  selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);
