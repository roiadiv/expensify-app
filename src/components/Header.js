import React from 'react';//using jsx
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {startLogout} from '../actions/auth';

export const Header = ({startLogout})=> (
    <header className="header">
      <div className = "content-container">
         <div className = "header__contant">
            <Link className="header__title" to = "/dashboard">
                <h1>Expense</h1>
            </Link>
         <button className = "button button--link" onClick = {startLogout}>Logout</button>
        </div>
      </div>
    </header>
);

const mapDispatchToProps = (dispatch)=>({
    startLogout:()=> dispatch(startLogout())   
});

export default connect(undefined, mapDispatchToProps)(Header);
