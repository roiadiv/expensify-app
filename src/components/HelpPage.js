import React from 'react';//using jsx

const HelpPage = ()=>(
    <div>
        <p>Help Page</p>
    </div>
);

export default HelpPage;