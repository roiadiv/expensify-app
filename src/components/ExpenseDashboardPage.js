import React from 'react';//using jsx
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage =  ()=>(
    <div>
        <ExpenseListFilters/>
        <ExpenseList />
    </div>
    
); 

export default ExpenseDashboardPage;