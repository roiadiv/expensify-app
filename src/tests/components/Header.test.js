import React from 'react';
import {shallow} from 'enzyme';
import {Header} from '../../components/Header';
// import ReactShallowRenderer from 'react-test-renderer/shallow'; we use enzyme

//We have to way to test our components
//1. shllow rendering - what gonna rendering
//2. full dom rendering - life cycle events and user actions

//For the header components we gonna use shallow rendering 

//we use Snapshot
//at the first time we run this code, Snapshot file will created
//the file have the data about the component

//at the aother time we run this code, the test will match the data inside the Snapshot file,
//and compre the changes inside the component file

//if thier is a changes the test warring us about us and ask if we want to update the changes,
//we make in the compenent file.
//for save - press 'u'
// after that the Snapshot file update

test('should render Header correct',()=>{
    //From using react-test-renderer
    // const renderer = new ReactShallowRenderer();
    // renderer.render(<Header />)
    // expect(renderer.getRenderOutput()).toMatchSnapshot();

    //To enzyme
    const wrapper = shallow(<Header  startLogout={()=>{}}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on button click',()=>{
    const startLogout = jest.fn();
    const wrapper = shallow(<Header startLogout = {startLogout} />);  
    wrapper.find('button').simulate('click');
    expect(startLogout).toHaveBeenCalled(); 
});