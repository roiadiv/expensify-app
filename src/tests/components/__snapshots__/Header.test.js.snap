// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correct 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__contant"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Expense
        </h1>
      </Link>
      <button
        className="button button--link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
